# Exam 1: Eduardo Abenza Severá

## Question 1

- Copy the **Locus_Germplasm_Phenotype_20130122.txt** file into your copy of the **Exams Git**
- Create a **new Jupyter Notebook** in your Exams folder called **"Exam_1_Answers"** using the 'bash' kernel
- Do the rest of this exam inside of the "Exam_1_Answers" Notebook

### Answer 1

Firstly, I'm creating a new folder called **Exam_1**, where I'm storing all the material I need to do this exam. I'm copying the **Locus_Germplasm_Phenotype_20130122.txt** file into **Exam_1 folder**, and I'm also moving my Jupyter Notebook **Exam_1_Answers** to this same folder. 

These are the Linux commands I'm using to do all these tasks:

In [5]:
mkdir Exam_1
cd Exam_1
cp ../Locus_Germplasm_Phenotype_20130122.txt Locus_Germplasm_Phenotype_20130122.txt
mv ../Exam_1_Answers.ipynb Exam_1_Answers.ipynb
echo "Contents of Exam_1 folder:"
echo ""
ls

Exam_1_Answers.ipynb  Locus_Germplasm_Phenotype_20130122.txt

Exam_1_Answers.ipynb notebook has been created.


## Question 2a

- Create a directory listing command that shows:
    - the **ownership** of the file
    - it's file-size, in **Megabytes** (i.e. human readable)
    
### Answer 2a

To show the **ownership** of **Locus_Germplasm_Phenotype_20130122.txt**, we use *ls -l* Linux command. In addition to that, we can include in this listing an extra *-h*, to indicate the size of this file in a **human-readable** way. The size of this file appears in **Mb**:

In [6]:
ls -lh Locus_Germplasm_Phenotype_20130122.txt

-rw-rw-r-- 1 osboxes osboxes 1.2M Sep  5 19:22 Locus_Germplasm_Phenotype_20130122.txt


## Question 2b

- Then, in words (in a Markdown box), describe the permissions on that file (read/write/execute) for users, groups, and "anyone"

### Answer 2b

Looking at the output of the previous code cell, we can state that:

- Users can **read and write** this file, but **they can't execute** it.
- Groups can **read and write** this file, but **they can't execute** it.
- "Anyone" can **read** this file, but **not write or execute** it.

## Question 3

- Create a command that outputs only the "header" line of Locus_Germplasm_Phenotype_20130122.txt

### Answer 3

To output only the header line of Locus_Germplasm_Phenotype_20130122.txt file (i.e., the first line), I'm using simply *head -1* Linux command.



In [7]:
head -1 Locus_Germplasm_Phenotype_20130122.txt

Locus_name)	Germplasm_name	phenotype	pubmed_id


In the previous code cell, we can see the name of the four columns in this file: **Locus_name)**, **Germplasm_name**, **phenotype** and **pubmed_id**.

## Question 4

- Create a command that outputs the total number of lines in Locus_Germplasm_Phenotype_20130122.txt

### Answer 4

In this piece of code, I'm calculating the total number of lines in this file. To do this in a clean manner, first I'm reading the content of the file using *cat* Linux command and then I'm piping this content into a line count (*wc -l*).

If I missed the first step of reading the file with *cat*, the file name would appear next to the line count and I don't like it that way.

In [12]:
cat Locus_Germplasm_Phenotype_20130122.txt | wc -l

7216


## Question 5a

- Create a command that writes ONLY the data lines (i.e. excludes the header!) to a new file called "Data_Only.csv"

### Answer 5a

As we saw in the previous question, this file has 7216 lines. The aim of this one is to exclude only the header, that is, the first line of the file.

This task can be achieved easily by using Linux command *tail -N*, which outputs the last N lines of a file. Taking into account that the header is the first line, it would be a nice idea to write the **last 7215 lines** of this file into *Data_Only.csv*.

In [13]:
tail -7215 Locus_Germplasm_Phenotype_20130122.txt > Data_Only.csv

## Question 5b

- Prove that your output file has the expected number of lines.

### Answer 5b
To prove that the output of Data_Only.csv file has the expected number of lines (this is, 7215 lines, 1 line shorter than the original file), we can obtain their total number of lines using *wc -l*.

This time, I have decided to include the file name next to its total number of lines, to provide further clarity.

In [19]:
wc -l Locus_Germplasm_Phenotype_20130122.txt
wc -l Data_Only.csv
echo ""
head -1 Data_Only.csv 

7216 Locus_Germplasm_Phenotype_20130122.txt
7215 Data_Only.csv

AT1G01040	CS3828	Increased abundance of miRNA precursors.	17369351


We can see that the file I have generated has exactly 7215 lines, which is one line shorter than the original document. We can also see the first line in *Data_Only.csv*, which shows that the one line missing is the header.

## Question 6

- Create a command that shows all of the lines that have a phenotype including the word "root"

### Answer 6

To complete this task, I'm using *grep* command to search for a specific pattern in Locus_Germplasm_Phenotype_20130122.txt. This pattern is the word *root*, either singular or plural, and in a case-insensitive way. As a matter of clarity, I am including the line number right before the matching line.

This is the code I'm using:

In [ ]:
grep -Ein 'root' Locus_Germplasm_Phenotype_20130122.txt 

## Question 7

- Create a command that writes the AGI Locus Code for every line that has a phenotype including the word "root" to a file called: Root-associated-Loci.txt

### Answer 7

In order to solve this exercise, I'm piping the output of the previous code cell into another grep command. In this second *grep* command, I'm looking for this pattern (indeed, this is the pattern indicated in the Exam 1 initial description):

- An A (located right after the beginning of the line) followed by a T.
    - Followed by one number between 1 and 5, both included.
        - Followed by a G, which in turn is followed by exactly five numbers.

In a case-insensitive way (*-i*). At the end, I **only** want the **AGI Loci Codes**, so I'm also including *-o* in the *grep* command.

Lastly, I'm writing this output in a new file called **Root-associated-Loci.txt**. Then:

In [ ]:
grep -Ei 'root' Locus_Germplasm_Phenotype_20130122.txt | grep -Eoi '^AT[1-5]G[0-9]{5}' > Root-associated-Loci.txt

## Question 8

- Create a command that writes the PubMed ID for every line that has a phenotype including the word "root" to a file called: Root-associated-Publications.txt

### Answer 8

The **PubMed ID** is located at the **end of the line**, and it is compound by a **variable length of numbers** between 0 and 9, both included. Then, first I'm using *grep* to look for all lines containing the word 'root', and then I search for this PubMed ID in the matching lines. In the final file **Root-associated-Publications.txt** I only want PubMed IDs, so I'm including *-o* in the second *grep* command.

In [ ]:
grep -Ei 'root' Locus_Germplasm_Phenotype_20130122.txt | grep -Eo '[0-9]+$' > Root-associated-Publications.txt

## Question 9

- Control experiment: You would hypothesize that genes associated with roots should be found on all chromosomes. Find a way (one or more commands) to test this hypothesis. In this dataset, is the hypothesis true? 

### Answer 9

To perform this control experiment, I'm using *grep* to look secuencially for AT1, AT2, AT3, AT4 and AT5 at the beginning of all lines in **Root-associated-Loci.txt**, in a case-insensitive way. 

In [31]:
echo "Chromosome 1"
grep -Eio '^[A][T]1' Root-associated-Loci.txt | wc -l
echo ""
echo "Chromosome 2"
grep -Eio '^[A][T]2' Root-associated-Loci.txt | wc -l
echo ""
echo "Chromosome 3"
grep -Eio '^[A][T]3' Root-associated-Loci.txt | wc -l
echo ""
echo "Chromosome 4"
grep -Eio '^[A][T]4' Root-associated-Loci.txt | wc -l
echo ""
echo "Chromosome 5"
grep -Eio '^[A][T]5' Root-associated-Loci.txt | wc -l
echo ""

Chromosome 1
215

Chromosome 2
183

Chromosome 3
126

Chromosome 4
69

Chromosome 5
233



As we can see in the previous code cell, genes associated with roots are found on all chromosomes.

## Question 10

- If your control experiment shows genes on every chromosome, then you can skip this question! (you answered Problem 9 correctly!)
- If your control experiment shows genes only on one or two chromosomes, then you have to explain why... what could the problem be? (I told you specifically to be careful about this problem!)

### Answer 10

Root-associated genes are distributed among all chromosomes, so I'm skipping this question.

## Question 11

- 'git commit' and 'git push' your answers to your GitHub, then give me your GitHub username before you leave the class. I will clone your repositories and grade your answers.

### Answer 11

In [ ]:
git commit -m "These are my answers for Exam 1."
git push